sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -1540,17 +1540,22 @@ device_image_plain ProgramManager::build(const device_image_plain &DeviceImage,
     const std::vector<unsigned char> &SpecConstsBlob =
         InputImpl->get_spec_const_blob_ref();
 
-    std::vector<device_image_impl::SpecConstDescT> &SpecConstOffsets =
-        InputImpl->get_spec_const_offsets_ref();
-
-    unsigned int PrevOffset = 0;
-    for (const device_image_impl::SpecConstDescT &SpecIDDesc :
-         SpecConstOffsets) {
-
-      Plugin.call<PiApiKind::piextProgramSetSpecializationConstant>(
-          NativePrg, SpecIDDesc.ID, SpecIDDesc.Offset - PrevOffset,
-          SpecConstsBlob.data() + SpecIDDesc.Offset);
-      PrevOffset = SpecIDDesc.Offset;
+    {
+      std::lock_guard<std::mutex> Lock{InputImpl->get_spec_const_data_lock()};
+      const std::map<std::string,
+                     std::vector<device_image_impl::SpecConstDescT>>
+          &SpecConstData = InputImpl->get_spec_const_data_ref();
+
+      for (const auto &DescPair : SpecConstData) {
+        for (const device_image_impl::SpecConstDescT &SpecIDDesc :
+             DescPair.second) {
+          if (SpecIDDesc.IsSet) {
```

Comment:
AFAIU, if not set, then the default value stored in memory must be set in case of SYCL2020. Please add a TODO.

Reply:
If the value is not set in RT, the default value determined during compilation will be used. So, it's redundant to set default values here.

Reply:
@kbobrovs I double checked with @AlexeySachkov, in native mode we don't have to set default value for spec constants, the frontend will extract it and store in SPIR-V (see https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#OpSpecConstant). And when spec constants are emulated, they will come in a separate buffer anyway.

Reply:
> If the value is not set in RT, the default value determined during compilation will be used.

that's how it worked with old spec constants, with 2020, default value specified by the user must be used.

Reply:
> If the value is not set in RT, the default value determined during compilation will be used.
> that's how it worked with old spec constants, with 2020, default value specified by the user must be used.

That's correct, but that default value is encoded by the compiler directly into generated SPIR-V: there is no need in additional handling from RT side